ERCS/erc-7546.md

```diff
@@ -0,0 +1,187 @@
+4. **Storage Layout** that ensures no conflicts when shared among multiple implementations.
+
+
+## Motivation
+This standard caters to:
+
+1. Users seeking the factory pattern for cloning contracts with common, upgradeable implementations.
+2. Scenarios requiring simultaneous upgrades of cloned contracts.
+3. Desires for flexible, per-function upgradeability.
+
+
+## Specification
+> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### 1. Dictionary Contract
+This contract manages a mapping of function selectors to corresponding implementation addresses to cater to calls from the Proxy.
+
+#### 1.1 Storage & Events
+The Dictionary MUST maintain a mapping of function selectors to implementation addresses.
+
+Changes to this mapping SHOULD be communicated through an event (or log).
+```solidity
+event ImplementationUpgraded(bytes4 indexed functionSelector, address indexed implementation);
+```
+
+#### 1.2 Functions
+The Dictionary SHOULD provide functions to retrieve and set implementations.
+
+##### 1.2.1 `getImplementation`
+```solidity
+function getImplementation(bytes4 functionSelector) external view returns (address);
+```
+
+##### 1.2.2 `setImplementation`
+```solidity
+function setImplementation(bytes4 functionSelector, address implementation) external;
+```
+The implementation contract MUST implement the function selector to be registered as a key in the mapping (section 3.1). Therefore, it is RECOMMENDED to verify this during registration.
+
+### 2. Proxy Contract
+This contract checks the Dictionary for the associated implementation address based on its function selector and delegatecalls to it.
+
+#### 2.1 Storage & Events
+The Proxy MUST reference the Dictionary address while ensuring a minimal chance of storage clashes. This is because the Dictionary's registered implementations will use this shared storage.
+
+Changes to the Dictionary address SHOULD be noticed:
+```solidity
+event DictionaryUpgraded(address indexed dictionary);
+```
+
+While it is possible to store the Dictionary address in the code area (e.g., using Solidity's immutable or constant), it SHOULD be designed with caution, considering the possibility that if the Dictionary's admin is not the same as the Proxy's admin, the ability to manipulate the implementation could be permanently lost.
+
+#### 2.2 Functions
+All calls to the Proxy MUST be forwarded to the implementation address registered in the Dictionary through the fallback function. Furthermore, as there is a possibility of collisions with function selectors registered in the Dictionary, Proxy SHOULD NOT have external functions.
+
+##### 2.2.1 `constructor`
+It is RECOMMENDED to initialize contracts, other than setting up the Dictionary address, through the Dictionary as well.
```

Comment:
What does "initialize contracts" mean here?

Reply:
We intended to recommend the use of an initialize function as a substitute for a constructor in the Proxy pattern. However, realizing this section could be left to the discretion of the implementation, we have decided to remove it.